File: leetcode/explore/binary tree/binary tree inorder traversal.cpp
#include <vector>
#include <stack>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> result;
        stack<TreeNode*> nodeStack;
        auto curNode = root;
        while (curNode != NULL || !nodeStack.empty())
        {
            while (curNode != NULL)
            {
                nodeStack.push(curNode);
                curNode = curNode->left;
            }
            curNode = nodeStack.top();
            nodeStack.pop();
            result.push_back(curNode->val);
            curNode = curNode->right;
        }
        return result;
    }

    vector<int> inorderTraversalRecurvise(TreeNode* root) {
        vector<int> result;
        recursiveHelper(root, result);
        return result;
    }

    void recursiveHelper(TreeNode* root, vector<int>& result) {
        if (!root) return;

        recursiveHelper(root->left, result);
        result.push_back(root->val);
        recursiveHelper(root->right, result);
    }
};

File: leetcode/explore/binary tree/binary tree inorder traversal.py
from typing import List


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        nodeStack = []
        result = []
        curNode = root
        while curNode or nodeStack:
            while curNode:
                nodeStack.append(curNode)
                curNode = curNode.left
            curNode = nodeStack.pop()
            result.append(curNode.val)
            curNode = curNode.right
        return result

    def inorderRecursive(self, root: TreeNode) -> List[int]:
        if not root: return []

        res = self.inorderRecursive(root.left)
        res.append(root.val)
        res += self.inorderRecursive(root.right)

        return res



if __name__ == "__main__":
    tn = TreeNode(1)
    root = tn
    tn.right = TreeNode(2)
    tn = tn.right
    tn.left = TreeNode(3)
    print(Solution().inorderTraversal(root))
    print(Solution().inorderRecursive(root))